netanalysis/ooni/data/ooni_client.py

```diff
@@ -17,69 +17,228 @@
-            Bucket=self._bucket,
-            Delimiter='/',
-            Prefix=self._prefix,
-            StartAfter=f'{self._prefix}{first_date.strftime("%Y%m%d")}',
-        )
+        if last_date < dt.date(2020, 10, 20):
+            return
+        paginator = self._s3_client.get_paginator('list_objects_v2')
+        pages = paginator.paginate(Bucket=NewOoniClient._BUCKET, Delimiter='/', Prefix=NewOoniClient._PREFIX,
+                                   StartAfter=f'{NewOoniClient._PREFIX}{first_date.strftime("%Y%m%d")}')
         for page in pages:
-            for date_entry in page.get('CommonPrefixes', []):
-                date_str = posixpath.basename(posixpath.dirname(date_entry['Prefix']))
+            self.num_list_requests += 1
+            for entry in page.get('CommonPrefixes', []):
+                date_dir = entry['Prefix']
+                date_str = posixpath.basename(posixpath.dirname(date_dir))
                 date = dt.datetime.strptime(date_str, "%Y%m%d").date()
                 if date > last_date:
                     return
                 for hour in range(24):
-                    prefix = f'''{date_entry['Prefix']}{hour:02}/{country}/'''
+                    prefix = f'''{date_dir}{hour:02}/{country}/'''
                     if test_type:
                         prefix += f'{test_type}/'
                     for page in paginator.paginate(Bucket=page['Name'], Prefix=prefix):
+                        self.num_list_requests += 1
                         for entry in page.get('Contents', []):
                             key = entry['Key']
-                            # Remove prefix
                             file_path = PosixPath(key)
                             if file_path.name.endswith('.jsonl.gz'):
                                 file_test_type = file_path.parent.name
-                                url = SplitResult('s3', self._bucket, key, None, None)
-                                yield FileEntry(self._client, file_test_type, country, date, url, entry['Size'])
-    
-    def _get_file(self, filename: PosixPath):
-        key = f'{self._prefix}{filename}'
-        return self._client.get_object(Bucket=self._bucket, Key=key)['Body']
\ No newline at end of file
+                                url = SplitResult(
+                                    's3', page['Name'], key, None, None)
+                                yield FileEntry(lambda: self._get_measurements(url), file_test_type, country, date, url, entry['Size'])
+
+    def _get_measurements(self, url: SplitResult) -> Iterable[Dict]:
+        s3_object = self._s3_client.get_object(Bucket=url.netloc, Key=url.path)
+        self.num_get_requests += 1
+        self.bytes_downloaded += s3_object['ContentLength']
+        with closing(s3_object['Body']) as remote_file, gzip.GzipFile(fileobj=remote_file, mode='r') as source_file:
+            for line in source_file:
+                yield ujson.loads(line)
+
+
+class LegacyOoniClient:
```

Comment:
Not a requirement, but should this and NewOoniClient inherit from a common interface?